### PR TITLE
add updateStrategy to daemonSets

### PR DESCRIPTION
**Pull request by hj-johannes-lee**
closes: #1500

### CODE REVIEW

**Comment by mythi** [comment]
> Please review though it is a draft yet!

I'd be very careful adding this to the API. Do you think it's useful? Our priority is to make our operator 'full lifecycle' compliant and if adding safe defaults to `deployments/*_plugin/base/intel-*-plugin.yaml` is sufficient that's what we should start with.

**Comment by hj-johannes-lee** [comment]
@mythi I think I still have no idea what makes our operator 'full cycle'..

Umm, the point that I do not understand is:
How can `deployments/*_plugin/base/intel-*-plugin.yaml` which has nothing to do with operator be related to making our operator full cycle?
I thought what we need to change is the operator part...

**Comment by mythi** [comment]
> How can `deployments/*_plugin/base/intel-*-plugin.yaml` which has nothing to do with operator

the operator uses these as the "skeleton" so changes there will add to operator automatically. See https://github.com/intel/intel-device-plugins-for-kubernetes/blob/main/deployments/daemonsets.go and https://github.com/intel/intel-device-plugins-for-kubernetes/blob/976466575bba32d4a915272fe1235be38b96b1b6/pkg/controllers/qat/controller.go#L83

**Comment by mythi** [comment]
> Ah do you mean... I did this
> 
> > define user configurable updateStrategy settings and add to the operator (to be decided/defined based on the need)
> 
> but didn't do this?
> 
> > enable default strategy in plugin daemonSets

yes, and the first part needs more thinking if we really want/need to do it. we cannot let users to set `[rollingUpdate.maxSurge](https://kubernetes.io/docs/reference/kubernetes-api/workload-resources/daemon-set-v1/#DaemonSetSpec)` since only 0 works for the plugins so that'd leave `[rollingUpdate.maxUnavailable](https://kubernetes.io/docs/reference/kubernetes-api/workload-resources/daemon-set-v1/#DaemonSetSpec)` AFAUI.

**Comment by hj-johannes-lee** [comment]
True. maxSurge should be always 0. Now I see your point.!

Then, if we can make users set only maxUnavailable, would it be good? Or, would we not need it since device plugins have already maxUnavailable value as default value (1)?

**Comment by mythi** [comment]
> Then, if we can make users set only maxUnavailable, would it be good?

my feedback is we don't need it configurable right now but let's prioritize the rest of the capability criteria

**Comment by hj-johannes-lee** [comment]
sure, then, let me just add other plugins' yamls with `controller_test.go` updated.

**Comment by hj-johannes-lee** [comment]
@mythi anything else to do?

File: pkg/controllers/dlb/controller_test.go [hunk 1]
```diff
@@ -34,6 +35,10 @@ func (c *controller) newDaemonSetExpected(rawObj client.Object) *apps.DaemonSet
 	devicePlugin := rawObj.(*devicepluginv1.DlbDevicePlugin)
 	yes := true
 	no := false
+	maxUnavailable := &intstr.IntOrString{}
+	maxUnavailable.IntVal = 1
+	maxSurge := &intstr.IntOrString{}
+	maxSurge.IntVal = 0
```
Comment (mythi) [inline pkg/controllers/dlb/controller_test.go:41]:
```suggestion
        maxUnavailable := intstr.FromInt(1)
        maxSurge := intstr.FromInt(0)
```

File: pkg/controllers/dlb/controller_test.go [hunk 2]
```diff
@@ -53,6 +58,13 @@ func (c *controller) newDaemonSetExpected(rawObj client.Object) *apps.DaemonSet
 					"app": appLabel,
 				},
 			},
+			UpdateStrategy: apps.DaemonSetUpdateStrategy{
+				Type: "RollingUpdate",
+				RollingUpdate: &apps.RollingUpdateDaemonSet{
+					MaxUnavailable: maxUnavailable,
+					MaxSurge:       maxSurge,
```
Comment (mythi) [inline pkg/controllers/dlb/controller_test.go:65]:
```suggestion
					MaxUnavailable: &maxUnavailable,
					MaxSurge:       &maxSurge,
```

**Review by mythi — COMMENTED** [review]
The values look fine. I hope this is what the operator sdk capability expects. One minor suggestion.

One other thing: I think your repo fork is broken due to the repo incident we had. Maybe you need to create a new fork and re-create the PR from there... let me check on this

**Pull request closed without merging**